dev/com.ibm.ws.microprofile.reactive.messaging.kafka/src/com/ibm/ws/microprofile/reactive/messaging/kafka/CommittingPartitionTracker.java

```diff
@@ -0,0 +1,304 @@
+ */
+public class CommittingPartitionTracker extends PartitionTracker {
+
+    private static final TraceComponent tc = Tr.register(CommittingPartitionTracker.class);
+
+    private final KafkaAdapterFactory factory;
+    private final ScheduledExecutorService executor;
+    private final KafkaInput<?, ?> kafkaInput;
+    private final int maxCommitBatchSize;
+    private final Duration maxCommitBatchInterval;
+
+    /**
+     * Set of CompletedWork which has either not been committed or has been committed but the commit has not completed yet.
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private final SortedSet<CompletedWork> completedWork = new TreeSet<>();
+
+    /**
+     * The count of CompletedWork for which a commit has not been started
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private int outstandingUncommittedWork = 0;
+
+    /**
+     * The last offset for which a commit was started
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private long committedOffset;
+
+    /**
+     * The currently scheduled task which will attempt to commit completed work or {@code null} if no task has been scheduled.
+     * <p>
+     * Any access to this must be synchronized
+     */
+    private Future<?> pendingCommitTask = null;
+
+    /**
+     *
+     * @param topicPartition the partition to track
+     * @param factory the KafkaAdaptorFactory
+     * @param kafkaInput the KafkaInput
+     * @param initialCommittedOffset the position of the reader when the partition was assigned
+     * @param executor a ScheduledExecutorService
+     * @param maxCommitBatchSize the maximum number of records to wait for before committing the offset
+     * @param maxCommitBatchInterval the maximum time to wait after a {@link #recordDone(long, Optional)} is called before committing the offset
+     */
+    public CommittingPartitionTracker(TopicPartition topicPartition,
+                                      KafkaAdapterFactory factory,
+                                      KafkaInput<?, ?> kafkaInput,
+                                      long initialCommittedOffset,
+                                      ScheduledExecutorService executor,
+                                      int maxCommitBatchSize,
+                                      Duration maxCommitBatchInterval) {
+        super(topicPartition);
+        this.factory = factory;
+        this.kafkaInput = kafkaInput;
+        this.executor = executor;
+        this.committedOffset = initialCommittedOffset;
+        this.maxCommitBatchSize = maxCommitBatchSize;
+        this.maxCommitBatchInterval = maxCommitBatchInterval;
+    }
+
+    @Override
+    public CompletionStage<Void> recordDone(long offset, Optional<Integer> leaderEpoch) {
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        synchronized (completedWork) {
+            completedWork.add(new CompletedWork(offset, leaderEpoch, result));
```

Comment:
It might never happen (or matter) but... you seem to have code that copes with the same offset being recorded more than once; if that happens should you really increment `outstandingUncommittedWork`?

Reply:
We ensure that everything works if `commitCompletedWork()` is called a second time before the `CompletionStage` from the previous commit has completed, but we don't actually do anything to protected against `recordDone()` being called twice with the same offset.